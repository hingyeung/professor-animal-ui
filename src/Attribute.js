import React, { Component } from 'react';

class Attribute extends Component {
  constructor(props) {
    super(props);

    this.name = props.name;
  }

  render() {
    return (
      <div class="attribute-row">
        <div class="attribute-cell">
          {this.name}
        </div>
        <div class="attribute-cell">
          <label htmlFor={this.name + '-yes'}>YES</label>
          <input type="radio" name={this.name} value="yes" id={this.name + '-yes'}/>
        </div>
        <div class="attribute-cell">
          <label htmlFor={this.name + '-no'}>NO</label>
          <input type="radio" name={this.name} value="no" id={this.name + '-no'}/>
        </div>
      </div>
    )
  }
}

export default Attribute;