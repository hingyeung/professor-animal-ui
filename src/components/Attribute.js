import React, {Component} from 'react';

class Attribute extends Component {
  constructor(props) {
    super(props);

    this.props = props;
    this.attribute = props.attribute;
    this.type = props.type;
    this.onAttributeChange = props.onAttributeChange;
  }

  onChange(attrType, attrName, e) {
    this.onAttributeChange(attrType, attrName, e.target.value);
  }

  render() {
    const yesId = this.type + '-' + this.attribute.name + '-yes',
          noId = this.type + '-' + this.attribute.name + '-no';

    return (
      <tr className="attribute-row row">
        <td className="attribute-cell col-sm-6">
          { this.attribute.name }
        </td>
        <td className="attribute-cell col-sm-3">
          <input className="form-check-label" type="radio" name={ this.type + '-' + this.attribute.name } value="yes"
                 id={ yesId }
                 onChange={ (e) => this.onChange(this.type, this.attribute.name, e) }
                 defaultChecked={this.attribute.value}/>
          <label className="form-check-label" htmlFor={ yesId }>YES</label>
        </td>
        <td className="attribute-cell col-sm-3">
          <input className="form-check-label" type="radio" name={ this.type + '-' + this.attribute.name } value="no"
                 id={ noId } onChange={ (e) => this.onChange(this.type, this.attribute.name, e) }
                 defaultChecked={!this.attribute.value}/>
          <label className="form-check-label" htmlFor={ noId }>NO</label>
        </td>
      </tr>
    )
  }
}

export default Attribute;