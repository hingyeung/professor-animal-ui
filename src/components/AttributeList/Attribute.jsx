import React from 'react';

function Attribute(props) {
  const attribute = props.attribute;
  const type = props.type;
  const onAttributeChange = props.onAttributeChange;
  const yesId = props.type + '-' + props.attribute.name + '-yes',
    noId = props.type + '-' + props.attribute.name + '-no',
    // Initially this.state.attributeValue is undefined (before animals are loaded),
    // and it makes the radio button "uncontrolled", then later when this.state.attributeValue
    // is populated with either true or false and the radio button becomes "controlled".
    // ReactJS doesn't like it and would give warning. Making sure its value is either true
    // or false prevents the problem.
    // https://github.com/facebook/react/issues/6779#issuecomment-222162404
    yesIsChecked = props.attribute.value ? props.attribute.value : false;

  function onChange(e) {
    onAttributeChange(type, attribute.name, e.target.value);
  }

  return (
    <tr className="attribute-row row">
      <td className="attribute-cell col-sm-6">
        { attribute.name }
      </td>
      <td className="attribute-cell col-sm-3">
        <input className="form-check-label" type="radio" name={ type + '-' + attribute.name } value="yes"
               id={ yesId }
               onChange={ onChange }
               checked={ yesIsChecked }/>
        <label className="form-check-label" htmlFor={ yesId }>YES</label>
      </td>
      <td className="attribute-cell col-sm-3">
        <input className="form-check-label" type="radio" name={ type + '-' + attribute.name } value="no"
               id={ noId } onChange={ onChange }
               checked={ !yesIsChecked }/>
        <label className="form-check-label" htmlFor={ noId }>NO</label>
      </td>
    </tr>
  )
}

export default Attribute;