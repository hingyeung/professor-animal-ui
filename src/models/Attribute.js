class Attribute {
  constructor(name, defaultValue) {
    this._name = name;
    this._defaultValue = defaultValue;
    this._value = '';
  }

  get name() {
    return this._name;
  }

  get defaultValue() {
    return this._defaultValue;
  };

  get value() {
    return this._value;
  }

  set value(value) {
    this._value = value;
  }
}

export default Attribute;