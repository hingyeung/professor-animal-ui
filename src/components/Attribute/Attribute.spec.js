import React from 'react';
import renderer from 'react-test-renderer';
import Attribute from "./Attribute";
import AttributeModel from 'models/Attribute';
import {shallow} from 'enzyme';

describe('Attribute', function () {
  const ATTRIBUTE_MODEL = new AttributeModel('name', true);
  const ATTRIBUTE_TYPE = 'type';
  const MOCK_ON_ATTRIBUTE_CHANGE_CB = jest.fn();
  beforeEach(function () {
    MOCK_ON_ATTRIBUTE_CHANGE_CB.mockClear();
  });

  it('should render properly', function () {
    const tree = renderer
    .create(<Attribute attribute={ ATTRIBUTE_MODEL } type={ ATTRIBUTE_TYPE }
                       onAttributeChange={ MOCK_ON_ATTRIBUTE_CHANGE_CB }/>)
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should call onAttributeChange() in props when attribute value changes', function () {
    const wrapper = shallow(<Attribute attribute={ ATTRIBUTE_MODEL } type={ ATTRIBUTE_TYPE }
                                       onAttributeChange={ MOCK_ON_ATTRIBUTE_CHANGE_CB }/>);
    const noRadioButton = wrapper.find('input[value="no"]');
    const event = {target: {value: 'no'}};
    noRadioButton.simulate('change', event);
    expect(MOCK_ON_ATTRIBUTE_CHANGE_CB).toBeCalledWith('type', 'name', 'no');
  })
});