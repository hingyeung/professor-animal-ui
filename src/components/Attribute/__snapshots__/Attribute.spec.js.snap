// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attribute should render properly 1`] = `
<tr
  className="attribute-row row"
>
  <td
    className="attribute-cell attribute-name col-sm-6"
  >
    name
  </td>
  <td
    className="attribute-cell col-sm-3"
  >
    <input
      checked={true}
      className="form-check-label attribute-value-yes"
      id="type-name-yes"
      name="type-name"
      onChange={[Function]}
      type="radio"
      value="yes"
    />
    <label
      className="form-check-label"
      htmlFor="type-name-yes"
    >
      YES
    </label>
  </td>
  <td
    className="attribute-cell col-sm-3"
  >
    <input
      checked={false}
      className="form-check-label attribute-value-no"
      id="type-name-no"
      name="type-name"
      onChange={[Function]}
      type="radio"
      value="no"
    />
    <label
      className="form-check-label"
      htmlFor="type-name-no"
    >
      NO
    </label>
  </td>
</tr>
`;
